### PR TITLE
CASEFLOW-1118 Submit Granted Substitution 

### PR DESCRIPTION
**Pull request by yoomlam**
Resolves [CASEFLOW-1118](https://vajira.max.gov/browse/CASEFLOW-1118) and bumps [CASEFLOW-1119](https://vajira.max.gov/browse/CASEFLOW-1119)

As an Intake or Clerk of the Board user, I want to see a success message when the appellant information is changed, so that I know the substitution was successful. 

### Description
Added controller to create an Appellant Substitution and a resulting new Appeal associated with a dispatched appeal.

### Acceptance Criteria
When the user selects "Confirm" the following occurs:
- [x] FE (frontend): Submit button works (sends payload to controller) 
- [x] FE (frontend): data (from Redux) is submitted to API
- [x] If no tasks selected the case is sent to case distribution
   * blocked by Task selection page
   * Reminder to write feature test once ready
- [x] FE: Success banner displays "You have successfully added a substitute appellant (Any necessary tasks have been created and assigned for further processing.)"  
- [ ] New appellant & POA information (if applicable) displays on the about the appellant and POA sections of the Case Details page
   * Relies on FE sending poa_participant_id as part of payload
   * From Ann-Marie: "For recognized appellants we are no longer going to be collecting the POA information manually. We will be pulling it from BGS as we are now. See updated ticket https://vajira.max.gov/browse/CASEFLOW-1044. We will introduce collecting the POA information when we implement substitution for unrecognized appellants"
- [x] New appeal stream is created
- [X] FE: User is redirected to Case Details page for new appeal stream
- [x] All issues on original appeal are on the new appeal
- [ ] Feature Toggle: recognized_granted_substitution_after_dd
   * controlled by FE; toggle not needed

### Testing Plan
Mimic testing done in spec/feature/queue/substitute_appellant/shared_setup.rb

### User Facing Changes
 - [ ] Screenshots of UI changes

BEFORE|AFTER
 ---|---
none | image

### Code Documentation Updates
- [x] Add or update code comments at the top of the class, module, and/or component.

### CODE REVIEW

File: client/app/queue/substituteAppellant/SubstituteAppellantContainer.jsx [hunk 1]
```diff
@@ -12,9 +12,9 @@ import { SubstituteAppellantReviewContainer } from './review/SubstituteAppellant
 
 const sections = [
   '1. Select substitute appellant',
-  '2. Select POA',
-  '3. Create task',
-  '4. Review',
+  // '2. Select POA',
```
Comment (yoomlam) [inline client/app/queue/substituteAppellant/SubstituteAppellantContainer.jsx:15]: https://dsva.slack.com/archives/CJL810329/p1618496348397500?thread_ts=1618495843.396300&cid=CJL810329

File: client/app/queue/substituteAppellant/basics/SubstituteAppellantBasicsView.jsx [hunk 2]
```diff
@@ -49,7 +49,7 @@ export const SubstituteAppellantBasicsView = () => {
     // Advance progressbar
     dispatch(stepForward());
 
-    history.push(`/queue/appeals/${appealId}/substitute_appellant/poa`);
+    history.push(`/queue/appeals/${appealId}/substitute_appellant/tasks`);
```
Comment (yoomlam) [inline client/app/queue/substituteAppellant/basics/SubstituteAppellantBasicsView.jsx:52]: Skipping Step 2: Select POA for now. Keeping `SubstituteAppellantPoaView` and `SubstituteAppellantPoaForm` for unrecognized appellants.

File: spec/controllers/cavc_remands_controller_spec.rb [hunk 3]
```diff
@@ -25,7 +25,7 @@
   end
 
   describe "POST /appeals/:appeal_id/cavc_remands" do
-    let(:source_appeal) { create(:appeal) }
+    let(:source_appeal) { create(:appeal, :dispatched) }
```
Comment (yoomlam) [inline spec/controllers/cavc_remands_controller_spec.rb:28]: Should only be applicable to dispatched appeals. Note that we don't validate that -- the frontend won't show the "Create CAVC Remand" button unless the appeal is dispatched.

File: app/controllers/appellant_substitutions_controller.rb [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    :source_appeal_id,
+    :substitution_date,
+    :claimant_type,
+    :substitute_participant_id,
+    :poa_participant_id,
+    :created_by_id
+  ].freeze
+
+  def create
+    new_gs = AppellantSubstitution.create!(create_params)
```
Comment (jcq) [inline app/controllers/appellant_substitutions_controller.rb:19]: NBD whatsoever, but "gs" might be confusing in the future to new folks who don't know "granted substitution" as a term

File: app/models/appeal.rb [hunk 5]
```diff
@@ -122,6 +122,9 @@ def create_stream(stream_type, new_claimants: nil)
       )).tap do |stream|
         if new_claimants
           new_claimants.each { |claimant| claimant.update(decision_review: stream) }
+
+          # Why isn't this a calculated value instead of stored in the DB?
```
Comment (jcq) [inline app/models/appeal.rb:126]: Good question... 🤔 

Perhaps overly zealous performance optimization?

File: app/models/appellant_substitution.rb [hunk 6]
```diff
@@ -8,7 +8,8 @@ class AppellantSubstitution < CaseflowRecord
   belongs_to :target_appeal, class_name: "Appeal"
 
   validates :created_by, :source_appeal, :substitution_date,
-            :claimant_type, :substitute_participant_id,
+            :claimant_type, # Claimant record type for the substitute
+            :substitute_participant_id,
             :poa_participant_id,
```
Comment (jcq) [inline app/models/appellant_substitution.rb:13]: Will this still pass if POA is empty?

File: app/controllers/appellant_substitutions_controller.rb (hunk 4 above)
Comment (yoomlam) [reply app/controllers/appellant_substitutions_controller.rb:19]: Changing to `new_substitution`.

File: app/models/appellant_substitution.rb (hunk 6 above)
Comment (yoomlam) [reply app/models/appellant_substitution.rb:13]: No, it is required. It's already required by the DB: https://github.com/department-of-veterans-affairs/caseflow/blob/8dc29c7eef3929a473d901457f974687f37f90b0/db/schema.rb#L131

Comment (yoomlam) [reply app/models/appellant_substitution.rb:13]: Is "No POA" an option?

Comment (yoomlam) [reply app/models/appellant_substitution.rb:13]: I think we can fix this and the DB schema once we understand the requirements for selecting a POA for unknown appellants.

**Review by jcq — APPROVED** [review]
Looks good! 

A few comments/suggestions, but the only thing that I'd consider a blocker is that bit about requiring POA, since I think it's likely that some appellants might not have one (?). That can likely be addressed later, and can be based on the answer to [this question](https://dsva.slack.com/archives/CJL810329/p1618523577442700).

File: client/app/queue/substituteAppellant/review/SubstituteAppellantReviewContainer.jsx [hunk 7]
```diff
@@ -29,12 +32,36 @@ export const SubstituteAppellantReviewContainer = () => {
...
-    // Redirect to Case Details page... but maybe for new appeal...?
-    history.push(`/queue/appeals/${appealId}`);
+      // Redirect to Case Details page... but maybe for new appeal...?
+      dispatch(
+        showSuccessMessage({
+          title: COPY.SUBSTITUTE_APPELLANT_SUCCESS_TITLE,
+          detail: COPY.SUBSTITUTE_APPELLANT_SUCCESS_DETAIL,
+        })
+      );
+      history.push(`/queue/appeals/${res.payload.target_appeal.uuid}`);
```
[the remark this answers is not in the document]
Comment (yoomlam) [reply client/app/queue/substituteAppellant/review/SubstituteAppellantReviewContainer.jsx]: Done

File: client/app/queue/substituteAppellant/substituteAppellant.slice.js [hunk 8]
```diff
@@ -74,15 +74,15 @@ export const completeSubstituteAppellant = createAsyncThunk(
...
-      const res = await ApiUtil.post('/appellant_substitutions', { data });
-      const attrs = res.body?.data?.attributes;
+      const res = await ApiUtil.post(`/appeals/${data.source_appeal_id}/appellant_substitution`, { data });
+      const attrs = res.body;
 
       return {
-        oldAppealId: attrs?.old_appeal_uuid,
-        newAppealId: attrs?.new_appeal_uuid,
+        substitution: attrs?.substitution,
+        target_appeal: attrs?.target_appeal,
```
[the remark this answers is not in the document]
Comment (yoomlam) [reply client/app/queue/substituteAppellant/substituteAppellant.slice.js]: Done

File: app/models/appellant_substitution.rb (hunk 6 above)
Comment (yoomlam) [reply app/models/appellant_substitution.rb:13]: Pushed changes to ignore poa_participant_id for now.

**Comment by yoomlam** [comment]
GitHub is showing prior commits that this branch was rebased on. This PR starts at commit `ce0466f`.

File: app/models/appellant_substitution.rb [hunk 9]
```diff
@@ -31,18 +33,49 @@ def power_of_attorney
...
 
-        InitialTasksFactory.new(target_appeal).create_root_and_sub_tasks!
+  def copy_request_issues(source_appeal, target_appeal)
+    source_appeal.request_issues.order(:id).map do |request_issue|
+      # This block of code may be a source of bugs as new columns are added to request_issues.
+      # It may be better to copy specific attributes, than duplicate everything.
+      request_issue.dup.tap do |request_issue_copy|
+        request_issue_copy.decision_review = target_appeal
+        # Do not copy decisions for new appeal
+        request_issue_copy.decision_date = nil
```
Comment (yoomlam) [inline app/models/appellant_substitution.rb:76]: Line 76 can be deleted as it is probably a bug due to my misunderstanding of `decision_date` at the time (I thought it was the decision date for the `request_issue.decision_issue`). @eileen-nava

**Pull request merged**